[hwasan] Optimize outlined memaccess for fixed shadow on Aarch64

thurstond opened the pull request:
The HWASan transform currently always uses x20 to pass the shadow base to hwasan_check_memaccess_shortgranules, even if the shadow base is a constant known at compile time (via -hwasan-mapping-offset). This patch introduces a fixed shadow variant of the hwasan_check_memaccess_shortgranules intrinsic, allowing the shadow base to be materialized inside the memaccess callee.

We currently only support this optimization for AArch64. It is a no-op on other platforms, or if -hwasan-mapping-offset is not specified.

Note: when -hwasan-mapping-offset is specified, it is necessary to specify HWASAN_OPTIONS=fixed_shadow_base=... (see ea991a11b2a3d2bfa545adbefb71cd17e8970a43) to ensure that the runtime will map the shadow appropriately.

thurstond commented on llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:
```diff
@@ -929,11 +929,33 @@ void HWAddressSanitizer::instrumentMemAccessOutline(Value *Ptr, bool IsWrite,
+  // shifted by 32 bits) - this is not an onerous constraint, since
+  // 1) kShadowBaseAlignment == 32 2) an offset of 4TB (1024 << 32) is
+  // representable, and ought to be good enough for anybody.
+  bool useFixedShadowIntrinsic = true;
```
Will do that

fmayer commented on llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:
```diff
@@ -625,14 +638,32 @@ void AArch64AsmPrinter::emitHwasanMemaccessSymbols(Module &M) {
-        *STI);
+
+    if (HwasanFixedShadowBase != (unsigned long long)-1) {
+      assert(IsShort);
```
(Implicit request: add this as a comment :))

thurstond replied:
Added comment (sans Polestar sponsored message)

vitalybuka replied:
At this stage better to make it simple without assumptions about android.
We have intrinsic then just lower it, it's up to HWASAN pass to generate code right.

fmayer replied:
The point of this comment is why we don't support non-short

vitalybuka replied:
We should support, instead of explaining.

And we may want non-shorts for SelHWASAN

vitalybuka reviewed (commented):
we need 2 tests:
1.  llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp
2.  llvm/lib/Target/AArch64/

If it's not a update to the test, please precommit the current behavior in a separate patch.

I'd would recommend to move HWAddressSanitizer.cpp and test into a patch
We can introduce intrinsic and lowering in the first patch.

I don't insist about splitting, though.

vitalybuka commented on llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:
```diff
@@ -929,11 +929,33 @@ void HWAddressSanitizer::instrumentMemAccessOutline(Value *Ptr, bool IsWrite,
+    IRB.CreateCall(Intrinsic::getDeclaration(
+                       M, UseShortGranules
+                              ? Intrinsic::hwasan_check_memaccess_shortgranules
+                              : Intrinsic::hwasan_check_memaccess),
```
Can we do `hwasan_check_memaccess_fixedshadow` for consistency?

thurstond replied:
I'll do this

thurstond replied:
Done

vitalybuka replied:
I don't see it here

vitalybuka replied:
Sorry, my request was to have both: short and no-short version for new intrinsics

thurstond replied:
Ok, will add both versions

vitalybuka commented on llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:
```diff
@@ -929,11 +929,33 @@ void HWAddressSanitizer::instrumentMemAccessOutline(Value *Ptr, bool IsWrite,
+  } else
+    useFixedShadowIntrinsic = false;
+
+  if (useFixedShadowIntrinsic)
```
Can we extract this logic into a function? so `if (useFixedShadowIntrinsic())`

fmayer replied:
I was thinking the same, but if you look at my other comment, I think once we make the logic shorter, it's not necessary.

vitalybuka commented on llvm/include/llvm/IR/Intrinsics.td:
```diff
@@ -2367,6 +2367,10 @@ def int_hwasan_check_memaccess :
 def int_hwasan_check_memaccess_shortgranules :
   Intrinsic<[], [llvm_ptr_ty, llvm_ptr_ty, llvm_i32_ty],
             [ImmArg<ArgIndex<2>>]>;
+def int_hwasan_check_memaccess_fixedshadow_shortgranules :
```
Can you document meaning of arguments.
I expected we are going to discard one of the arguments, like `int_asan_check_memaccess`.

thurstond replied:
I'll add docs.

The way ASan avoids the need to pass the shadow offset via the intrinsic is by
defining llvm::getAddressSanitizerParams() in lib/Transforms/Instrumentation/AddressSanitizer.cpp. I'll try adapting that to this patch.

thurstond replied:
Done

thurstond replied:
To discard the arguments, I ended up adding getFixedShadowBase() to the HWAddressSanitizer pass

vitalybuka commented on llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:
```diff
@@ -929,11 +929,33 @@ void HWAddressSanitizer::instrumentMemAccessOutline(Value *Ptr, bool IsWrite,
+  // 1) kShadowBaseAlignment == 32 2) an offset of 4TB (1024 << 32) is
+  // representable, and ought to be good enough for anybody.
+  bool useFixedShadowIntrinsic = true;
+  if (TargetTriple.isAArch64() && ClMappingOffset.getNumOccurrences() > 0 &&
```
We better do not touch ClMappingOffset any more.
We copied that into Mapping?

thurstond replied:
Mapping.Offset could be zero because of a manually specified ClMappingOffset option, or because it is Fuchsia or Khwasan. I want to make sure that this optimization is only applied for a manually specified ClMappingOffset option.

```
  if (TargetTriple.isOSFuchsia()) {
    ...
    Offset = 0;
  } else if (ClMappingOffset.getNumOccurrences() > 0) {
    ...  
    Offset = ClMappingOffset;
  } else if (ClEnableKhwasan || InstrumentWithCalls) {
    ...
    Offset = 0;
```

vitalybuka replied:
Why do we need to exclude them?

thurstond replied:
I'm concerned it may break things, because it will automatically opt Fuchsia and Khwasan into using the fixedshadow intrinsic (unlike on Aarch64 HWASan, where this patch only takes effect if they opt-in to fixed shadows). I don't have the infrastructure set up to test my change on Fuchsia or Khwasan.

vitalybuka replied:
This is not a reason to complicate with unnecessary branches. Unless you have a good reason, please make it as generic as possible.

thurstond replied:
A new reason: to avoid needing the shadow base in the intrinsic, I've added a getFixedShadowBase() function, but that is in the llvm namespace and doesn't have access to Mapping.

thurstond commented:
> we need 2 tests:
> 
> 1. llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp
> 2. llvm/lib/Target/AArch64/
> 
> If it's not a update to the test, please precommit the current behavior in a separate patch.
> 
> I'd would recommend to move HWAddressSanitizer.cpp and test into a patch We can introduce intrinsic and lowering in the first patch.
> 
> I don't insist about splitting, though.

Coming soon!

thurstond committed:
[hwasan] Optimize outlined memaccess for fixed shadow on Aarch64

The HWASan transform currently always uses x20 to pass the shadow base to hwasan_check_memaccess_shortgranules, even
if the shadow base is a constant known at compile time (via -hwasan-mapping-offset). This patch uses the fixed
shadow variant of the hwasan_check_memaccess_shortgranules intrinsic (introduced in
https://github.com/llvm/llvm-project/commit/365bddf634993d5ea357e9715d8aacd7ee40c4b5), allowing the shadow base to
be materialized inside the memaccess callee.

We currently only support this optimization for AArch64. It is a no-op on other platforms, or if -hwasan-mapping-offset is not specified.

Note: when -hwasan-mapping-offset is specified, it is necessary to specify HWASAN_OPTIONS=fixed_shadow_base=... (see ea991a1) to ensure that the runtime will map the shadow appropriately.

thurstond commented:
Rebased on top of the memaccess intrinsics patch

The test case for this CL already exists in llvm-project/compiler-rt/test/hwasan/TestCases/Linux/fixed-shadow.c

vitalybuka commented on llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:
```diff
@@ -930,11 +930,32 @@ void HWAddressSanitizer::instrumentMemAccessOutline(Value *Ptr, bool IsWrite,
+  // representable.
+  // In particular, an offset of 4TB (1024 << 32) is representable, and
+  // ought to be good enough for anybody.
+  if (TargetTriple.isAArch64() && ClMappingOffset.getNumOccurrences() > 0) {
```
Can we do `&& Mapping.Offset != kDynamicShadowSentinel` ?
so we can make Mapping as source of truth after `ShadowMapping::init`

thurstond replied:
Updated in https://github.com/llvm/llvm-project/pull/88544/commits/cdbecc6603e835cd9ff2b10fcdd1ba5bd467a1b6

vitalybuka commented on llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:
```diff
@@ -930,11 +930,32 @@ void HWAddressSanitizer::instrumentMemAccessOutline(Value *Ptr, bool IsWrite,
+    IRB.CreateCall(
+        Intrinsic::getDeclaration(
+            M, UseShortGranules
+                   ? Intrinsic::hwasan_check_memaccess_shortgranules_fixedshadow
```
We need a hwasan pass specific test

thurstond replied:
Updated fixed-shadow.ll test: https://github.com/llvm/llvm-project/pull/88544/commits/cdbecc6603e835cd9ff2b10fcdd1ba5bd467a1b6

vitalybuka commented on llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:
```diff
@@ -930,11 +930,32 @@ void HWAddressSanitizer::instrumentMemAccessOutline(Value *Ptr, bool IsWrite,
-                 {ShadowBase, Ptr, ConstantInt::get(Int32Ty, AccessInfo)});
+  bool useFixedShadowIntrinsic = false;
+  // The memaccess fixed shadow intrinsic is only supported on AArch64,
+  // which allows a 16-bit immediate to be left-shifted by 32.
```
what is `16-bit immediate` ?

vitalybuka replied:
I see, you want only particular constants to make lowering efficient

thurstond committed:
Update fixed-shadow.ll test; use 'Mapping.Offset != kDynamicShadowSentinel' per Vitaly's feedback

thurstond committed:
Fix formatting

thurstond committed:
Update two more existing tests
